src/lib.rs

```diff
@@ -876,6 +879,72 @@ impl Future for IpfsFuture {
                 }
             }
 
+            while let Poll::Ready(Some(evt)) = Pin::new(&mut self.bitswap_events).poll_next(ctx) {
+                match evt {
+                    BitswapEvent::ReceivedBlock(peer_id, block) => {
+                        let repo = self.repo.clone();
+                        let peer_stats = Arc::clone(
+                            &self
+                                .swarm
+                                .bitswap()
+                                .connected_peers
+                                .get(&peer_id)
+                                .unwrap()
+                                .stats,
+                        );
+                        task::spawn(async move {
+                            let bytes = block.data().len() as u64;
+                            let res = repo.put_block(block.clone()).await;
+                            match res {
+                                Ok((_, uniqueness)) => match uniqueness {
+                                    BlockPut::NewBlock => peer_stats.update_incoming_unique(bytes),
+                                    BlockPut::Existed => {
+                                        peer_stats.update_incoming_duplicate(bytes)
+                                    }
+                                },
+                                Err(e) => {
+                                    debug!(
+                                        "Got block {} from peer {} but failed to store it: {}",
+                                        block.cid,
+                                        peer_id.to_base58(),
+                                        e
+                                    );
+                                    return;
+                                }
+                            };
+                        });
+                    }
+                    BitswapEvent::ReceivedWant(peer_id, cid, priority) => {
+                        info!(
+                            "Peer {} wants block {} with priority {}",
+                            peer_id.to_base58(),
+                            cid.to_string(),
+                            priority
+                        );
+
+                        let repo = self.repo.clone();
+                        let queued_blocks = Arc::clone(&self.swarm.bitswap().queued_blocks);
+                        task::spawn(async move {
+                            let block = match repo.get_block(&cid).await {
```

Comment:
Unrelated to anything introduced in this PR, perhaps a byproduct of the changes is that using `get_block` directly will put to block on our wantlist. If we had working discovery, we would be gathering wantlists from every peer we meet and thus being highly helpful in the network while running out of memory to help everyone we've met :)

```suggestion
                            // FIXME: need a better API to `get_block_now` or without waiting.
                            // use of `get_block` here results in us trying wanting the block. while
                            // such "helping a friend out" might be useful that'd be an highly
                            // advanced feature we probably don't yet want to pursue.
                            let block = match repo.get_block(&cid).await {
```

Reply:
this will be handled on a follow-up PR with even more changes